demos/demo_stacking_simulation.py

```diff
@@ -0,0 +1,1143 @@
+		'''
+		if offset is not None:
+			self.object_size += offset
+
+		# Initial Values
+		i = 0
+		obj_start_pos = self.object_position
+
+		while True:
+			# Slowly changing the y-position for the end_effector
+			x = (1 - math.cos(math.radians(i/10))) * (desired_position[0] - obj_start_pos[0])/2 + obj_start_pos[0]
+			y = (1 - math.cos(math.radians(i/10))) * (desired_position[1] - obj_start_pos[1])/2 + obj_start_pos[1]
+			z = (1 - math.cos(math.radians(i/10))) * (desired_position[2] - obj_start_pos[2])/2 + obj_start_pos[2]
+
+			self.object_position = [x,y,z]
+
+			# Calculating desired_external_force to apply to grasp the object
+			if self.is_grabbing:
+				desired_external_force = self.get_desired_external_force()
+			
+			# Getting final state for end_effector
+			self.get_desired_state()
+			
+			# Get robot's observed state
+			self.get_next_state(debug=False)
+
+			# Calculating required torque for total force
+			self.get_torques()
+
+			# Setting timeout for "timeout seconds"
+			i += 1
+			if i % (timeout*1000) == 0:
+				print("****************** Done *******************")
+				break
+
+			if exit_before_timeout and all(np.abs(np.subtract(self.object_position, desired_position)) < 0.005):
+				break
+
+		if offset is not None:
+			self.object_size -= offset
+
+
+
+	def grasp(self, f_applied=[1.1,1.1,0]):
+		self.f_applied = f_applied
+		# self.is_grabbing = True
+		self.cnt_array = [1,1,1]
+		self.move(self.object_position, timeout=1, exit_before_timeout=False)
+		self.is_grabbing = True
+
+
+	def release(self):
+		self.is_grabbing = False
+		self.cnt_array = [0,0,0]
+		self.desired_external_force = np.matrix(np.zeros(9)).T
+
+
+	def follow_predefined_trajectory(self):
+		assert self.object_position is not None
+		assert self.object_size is not None
+		initial_position = self.object_position
+		self.object_position = [0,0,0]
+		self.move([0,0,0.25], offset=0.04, timeout=2, exit_before_timeout=False)
+		if self.mode == "simulation":
+			self.finger.set_block_state(initial_position, [0,0,0,1])
+		self.move(initial_position, offset=0.04, timeout=2, exit_before_timeout=False) # reaching the object
+		self.grasp()
+		self.move(self.object_position[:2] + [0.21]) # picking up 
+		self.move(self.landing_position[:2] + [self.object_position[2]]) # moving forward
+		self.move(self.landing_position) # moving down
+		self.release()
+		self.move(self.landing_position[:2] + [0.25], offset=0.02) # moving away from the object after placing it
+		self.move([0,0,0.25], offset=0.02) # moving to the centre (just using it for a better ending, for now)
+
+
+	def trajectory2(self):
+		self.move(self.object_position, offset=0.02, timeout=1, exit_before_timeout=False) # reaching the object
+		self.grasp()
+		for _ in range(5):
+			self.move(self.object_position[:2] + [0.21]) # picking up 
+			self.move(self.object_position[:2] + [0.045]) # picking up 
+
+
+	def next_state(self):
+		seconds = 6
+		if (self.t*self.dt) % seconds == 0:
```

Comment:
Using `a % b == 0` with `a` being a float (which is the case here due to `self.dt`) may fail in some cases due to float inaccuracy (classical example: `0.1 * 3 != 0.3`).

Reply:
A save alternative here would be `self.t % int(seconds / self.dt) == 0`